runtime/near-vm-logic/src/receipt_manager.rs

```diff
@@ -0,0 +1,493 @@
+    ///
+    /// # Arguments
+    ///
+    /// * `receipt_index` - an index of Receipt to append an action
+    /// * `public_key` - a public key for an access key to delete
+    ///
+    /// # Panics
+    ///
+    /// Panics if the `receipt_index` does not refer to a known receipt.
+    pub fn append_action_delete_key(
+        &mut self,
+        receipt_index: u64,
+        public_key: Vec<u8>,
+    ) -> ExtResult<()> {
+        self.append_action(
+            receipt_index,
+            Action::DeleteKey(DeleteKeyAction {
+                public_key: PublicKey::try_from_slice(&public_key)
+                    .map_err(|_| HostError::InvalidPublicKey)?,
+            }),
+        );
+        Ok(())
+    }
+
+    /// Attach the [`DeleteAccountAction`] action to an existing receipt
+    ///
+    /// # Arguments
+    ///
+    /// * `receipt_index` - an index of Receipt to append an action
+    /// * `beneficiary_id` - an account id to which the rest of the funds of the removed account will be transferred
+    ///
+    /// # Panics
+    ///
+    /// Panics if the `receipt_index` does not refer to a known receipt.
+    pub fn append_action_delete_account(
+        &mut self,
+        receipt_index: u64,
+        beneficiary_id: AccountId,
+    ) -> ExtResult<()> {
+        self.append_action(
+            receipt_index,
+            Action::DeleteAccount(DeleteAccountAction { beneficiary_id }),
+        );
+        Ok(())
+    }
+
+    /// Distribute the gas among the scheduled function calls that specify a gas weight.
+    ///
+    /// Distributes the gas passed in by splitting it among weights defined in `gas_weights`.
+    /// This will sum all weights, retrieve the gas per weight, then update each function
+    /// to add the respective amount of gas. Once all gas is distributed, the remainder of
+    /// the gas not assigned due to precision loss is added to the last function with a weight.
+    ///
+    /// # Arguments
+    ///
+    /// * `gas` - amount of unused gas to distribute
+    ///
+    /// # Returns
+    ///
+    /// Function returns a [GasDistribution] that indicates how the gas was distributed.
+    #[cfg(feature = "protocol_feature_function_call_weight")]
+    pub fn distribute_unused_gas(&mut self, gas: u64) -> GasDistribution {
+        let gas_weight_sum: u128 =
+            self.gas_weights.iter().map(|(_, GasWeight(weight))| *weight as u128).sum();
+        if gas_weight_sum != 0 {
+            // Floor division that will ensure gas allocated is <= gas to distribute
+            let gas_per_weight = (gas as u128 / gas_weight_sum) as u64;
+
+            let mut distribute_gas = |metadata: &FunctionCallActionIndex, assigned_gas: u64| {
+                let FunctionCallActionIndex { receipt_index, action_index } = metadata;
+                if let Some(Action::FunctionCall(FunctionCallAction { ref mut gas, .. })) = self
+                    .action_receipts
+                    .0
+                    .get_mut(*receipt_index)
```

Comment:
Let's extract this bit into `fn get_fuction_call_action_mut` function, it'll more clearl y explain the purpose of FunctionCallIndex construct.

Reply:
The reason I didn't do this is because this closure is used within code blocks where `self` is borrowed immutably, so this wouldn't be able to be used. I'll extract it into a helper function but I'm unsure that's better